Ton support

PolyProgrammist opened the pull request:
## Description

<!--- Describe your changes in detail -->

## How to test

<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] Create pull request as draft initially, unless its complete.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

stspbu committed:
Register TON and add its stub files

stspbu committed:
Add wallet implementation (only wallet V4R2 for now)
Add test for deriveAddress checking

stspbu committed:
Add user-friendly address calculations
Fix TON address tests accordingly

stspbu committed:
Add address validation and parsing
Address is inherited from Everscale::Address
Wallet returns bounceable address by default
Address tests finalized

stspbu committed:
Migrate on WalletV4R2 explicit usage and remove PublicKey constructor from Address
Add missing Address tests
Correct existing Address tests

stspbu committed:
Declare TON proto
Create TON Signer
Create TON Signer tests
Refactor Everscale Message to fit both chains
Unify and fix some imports

stspbu committed:
Fix code style & write type for workchainId

stspbu committed:
Add test transfers with comments
Remove corresponding TODO

stspbu committed:
Add ios tests
Fix implicit type casts

PolyProgrammist commented on src/Everscale/Messages.cpp:
```diff
@@ -130,15 +132,10 @@ Data createSignedMessage(PublicKey& publicKey, PrivateKey& key, bool bounce, uin
-    message.setBody(body);
-
-    const auto messageCell = message.intoCell();
+    message.setBody(payload.intoCell());
```
По какой причине нужно было здесь поменять?

stspbu replied:
Мы тип поменяли, теперь body это Cell, а не CellSlice
А нужно мне это для того, чтобы в конечном итоге мы передавали тело ссылкой (во-первых, так больше помещается, во-вторых, так делает tonweb, потому что toncenter пока не умеет правильно обрабатывать body как slice)

PolyProgrammist replied:
А зачем поменяли тип?

PolyProgrammist commented on src/proto/TheOpenNetwork.proto:
```diff
@@ -0,0 +1,67 @@
+    WalletV4R2 = 2;
+};
+
+enum SendMode {
```
Почему выбраны именно такие возможности?

stspbu replied:
Это все, что есть в документации:
https://ton.org/docs/develop/func/stdlib#send_raw_message

PolyProgrammist replied:
Не должно разве быть дополнительно к этому 33, 34, 65, 66, 96, и так далее? 
>mode' = mode + 1 means that the sender wants to pay transfer fees separately; mode' = mode + 2 means that any errors arising while processing this message during the action phase should be ignored. Finally, mode' = mode + 32

stspbu replied:
> Не должно разве быть дополнительно к этому 33, 34, 65, 66, 96, и так далее?
> 
> > mode' = mode + 1 means that the sender wants to pay transfer fees separately; mode' = mode + 2 means that any errors arising while processing this message during the action phase should be ignored. Finally, mode' = mode + 32

Как получить 34, имея 32 и 2? :)
=> DESTROY_ON_ZERO_BALANCE | IGNORE_ACTION_PHASE_ERRORS
Это классический сишный прикол, в тестах есть соответствующие примеры

PolyProgrammist commented on src/TheOpenNetwork/Signer.cpp:
```diff
@@ -0,0 +1,66 @@
+        break;
+    }
+    default:
+        throw std::invalid_argument("Unsupported wallet version");
```
std::runtime_error is more common for this type

stspbu replied:
> std::runtime_error is more common for this type

судя по примерам по коду охотнее используется invalid_argument, когда действительно invalid argument
вот, например:

```
if (publicKey.type != TWPublicKeyTypeED25519) {
    throw std::invalid_argument("address may only be an extended ED25519 public key");
}
```

PolyProgrammist commented on src/proto/TheOpenNetwork.proto:
```diff
@@ -0,0 +1,67 @@
+    uint32 mode = 5;
+
+    // Expiration UNIX timestamp (optional, now() + 60 by default)
+    uint32 expired_at = 6;
```
uint64

PolyProgrammist commented on src/proto/TheOpenNetwork.proto:
```diff
@@ -0,0 +1,67 @@
+    // Learn more: https://ton.org/docs/develop/smart-contracts/guidelines/external-messages
+    uint32 seqno = 4;
+
+    // Send mode (optional)
```
0 by default

stspbu replied:
там, где и так понятно, какой дефолт, я не писал его явно
int - 0, string - ""

PolyProgrammist replied:
Но в seqno 0 by default написано. Лучше чтобы везде было консистентно

stspbu replied:
> Но в seqno 0 by default написано. Лучше чтобы везде было консистентно

Там написано, для чего этот 0 используется, чтобы было понятно, что будет происходить, если он будет 0. Но ок, можно и везде написать. И в comment ты предлагаешь писать "" by default?

 Еще можно мою фразу в seqno переформулировать, например. С ходу не знаю правда, как.

PolyProgrammist replied:
Ну мне просто сходу непонятно какой дефолтный seqno. А коммент - не знаю, он как-то более понятен, но можно и там написать

PolyProgrammist commented on src/TheOpenNetwork/Address.h:
```diff
@@ -0,0 +1,57 @@
+    [[nodiscard]] std::string string(bool userFriendly, bool bounceable = true, bool testOnly = false) const;
+
+private:
+    static Data decodeUserFriendlyAddress(const std::string& string);
```
Это можно убрать отсюда и оставить только в cpp файле

stspbu replied:
> Это можно убрать отсюда и оставить только в cpp файле

Зачем? Имхо, лучше, чтобы было видно весь интерфейс в хедере

PolyProgrammist replied:
decodeUserFriendlyAddress к интерфейсу не относится, это детали реализации

stspbu replied:
> decodeUserFriendlyAddress к интерфейсу не относится, это детали реализации

Это деталь реализации, поэтому это приватный метод
Относится к внутреннему интерфейсу сущности Address

PolyProgrammist replied:
Он не нужен внешним пользователям класса Address. Поэтому из интерфейса (Address.h) его можно убрать. Тогда пользователи класса Address не будут зря задумываться о том, что это такое, и не будут зря компилировать эту строчку кода.

stspbu replied:
> Он не нужен внешним пользователям класса Address. Поэтому из интерфейса (Address.h) его можно убрать. Тогда пользователи класса Address не будут зря задумываться о том, что это такое, и не будут зря компилировать эту строчку кода.

По такой логике приватные методы вообще не нужны? :)

PolyProgrammist replied:
Они нужны когда используется сама сущность адреса. Иначе просто никак не сделаешь, если использовать приватные поля. А тут можно

PolyProgrammist commented on src/TheOpenNetwork/Address.cpp:
```diff
@@ -0,0 +1,128 @@
+    Data decoded;
+    try {
+        decoded = decodeUserFriendlyAddress(string);
+    } catch (...) {
```
Лучше ловить конкретные ошибки

stspbu replied:
> Лучше ловить конкретные ошибки

В общем случае лучше, но не в этом контексте. Здесь смысл в том, чтобы сделать этот метод noexcept, а не чтобы обрабатывать ошибки, см. другие примеры в коде

stspbu replied:
Примеры: 
wallet-core/src/Bitcoin/MessageSigner.cpp
wallet-core/src/Cardano/AddressV3.cpp

и другие в таком же ключе ищутся по проекту

PolyProgrammist commented on src/TheOpenNetwork/Signer.cpp:
```diff
@@ -0,0 +1,66 @@
+        try {
+            const auto& transferMessage = Signer::createTransferMessage(publicKey, privateKey, transfer);
+            protoOutput.set_encoded(TW::Base64::encode(transferMessage));
+        } catch (...) { }
```
Лучше ловить конкретные ошибки

stspbu replied:
Тоже самое

PolyProgrammist commented on src/TheOpenNetwork/Address.cpp:
```diff
@@ -0,0 +1,128 @@
+
+    byte tag = decoded[0];
+    if (tag & AddressTag::TEST_ONLY) {
+        tag ^= AddressTag::TEST_ONLY;
```
Почему так? Информация ни на что не влияет получается?

PolyProgrammist replied:
Точнее судя по следующей строке, с адресами тестнета нельзя работать. Почему так?

stspbu replied:
> Точнее судя по следующей строке, с адресами тестнета нельзя работать. Почему так?

Я не знаю, как у тебя так получилось, но смысл в том, что если есть тестовый флаг, мы его убираем
А дальше смотрим на остаток, который должен быть либо bounceable, либо нет

stspbu committed:
Remove non-hardened items from TON derivation path
Add android (kotlin) tests
Rename expiredAt => expireAt and make it uint64 in proto
Other minor code style fixes

PolyProgrammist commented on src/TheOpenNetwork/Message.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace TW::TheOpenNetwork {
+
+Cell::Ref Message::intoCell() const {
```
Тут по сути тоже самое что и в everscale. Можно ли как-то объединить код?

PolyProgrammist replied:
Типы только разные

stspbu replied:
> Типы только разные

Нет, не только, здесь еще другие биты пишутся, чтобы сигнализировать, что внутри рефка
Сейчас думаю, что не стоит извращаться и пытаться все объединить, не такой существенный копипаст тут

P.S. Изначально думал о том, чтобы сделать реализацию по рефке общей, но после этого придется менять вообще все тесты у everscale (потому что получится уже другой boc), это слишком трудозатратно и того не стоит

stspbu committed:
Apply Entry::deriveAddress() unification for TON (trustwallet#2796)

PolyProgrammist committed:
Move implementation method to .cpp file

stspbu committed:
Use std::move to escape unwanted copy in Wallet constructor
Add inline to decodeUserFriendlyAddress

stspbu committed:
Revert expire_at uint64 and make it uint32

PolyProgrammist committed:
improve order

PolyProgrammist committed:
Include chrono

PolyProgrammist committed:
Update include/TrustWalletCore/TWBlockchain.h

Co-authored-by: hewigovens <360470+hewigovens@users.noreply.github.com>

PolyProgrammist committed:
Review edits:
seqno -> sequence_number
upper case in WalletVersion enum
replaced mode with enum values in kotlin

PolyProgrammist committed:
enum fix test for android

PolyProgrammist committed:
Add tests for wasm

PolyProgrammist committed:
constants fixes

PolyProgrammist committed:
Fix exceptions in address validation: stoi and base64

PolyProgrammist committed:
Fix exception - check if address is hex encoded before parsing

PolyProgrammist committed:
Fix tests for everscale address

stspbu committed:
Rename coin to TON
Rename blockchain to TheOpenNetwork

stspbu committed:
Extend Address hash constructor with address type flags
Remove auto generated comments in entry.h/cpp files

PolyProgrammist committed:
Moved address to separate class to avoid inheritance

PolyProgrammist committed:
Moved message to separate class to avoid inheritance

PolyProgrammist committed:
Using error code for unsupported wallet version

stspbu committed:
Move other common parts to shared folder (Cell primitives, StateInit, etc)
Move out Everscale dependencies from CommonTON (e.g. Wallet-related InitData, createSigningMessage and so on)
Remove left Everscale dependencies from TheOpenNetwork (replaced with CommonTON)
Cleanup unused includes

optout21 committed:
Add extra hrp and coin check to lockScriptForAddress() (#2814)

AyushBherwani1998 committed:
Native Injective integration (#2828)

* add basic implementation of injective
* add swift tests
* add cpp tests

Milerius committed:
misc(CODEOWNERS): update CODEOWNERS (#2861)

Milerius committed:
feat(thorswap): SWAP from atom support (#2860)

stspbu commented:
Done

PolyProgrammist commented:
🔥
